File: leetcode/lib/217_contains_duplicate.rb
# Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

# use a hash map
# for num in nums
#		if num in hash: return true
#		else: add num to hash
# return false at end

def contains_duplicate(nums)
	dupes = Hash.new(0)

	for num in nums do
		if dupes.key?(num)
			return true
		else
			dupes[num] += 1
		end
	end

	return false
end

puts("--- TEST 1 ---")
nums = [1,2,3,1]
puts(contains_duplicates(nums))
# expect true

puts("--- TEST 2 ---")
nums = [1,2,3,4]
puts(contains_duplicates(nums))
# expect false

puts("--- TEST 3 ---")
nums = [1,1,1,3,3,4,3,2,4,2]
puts(contains_duplicates(nums))
# expect true

File: leetcode/lib/arrays_hashing/217_contains_duplicate.py
"""
Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.
"""

from utils import test


class Solution:
    def contains_duplicate(self, nums: list[int]) -> bool:
        duplicates = set()

        for num in nums:
            if num in duplicates:
                return True
            else:
                duplicates.add(num)
        return False


def main():
    solution = Solution()

    test(1, True, solution.contains_duplicate([1, 2, 3, 1]))
    test(2, False, solution.contains_duplicate([1, 2, 3, 4]))
    test(3, True, solution.contains_duplicate([1, 1, 1, 3, 3, 4, 3, 2, 4, 2]))


if __name__ == "__main__":
    main()
